### PR TITLE
Add custom signal handling probe type

### PR DESCRIPTION
**Pull request by amiremohamadi**
#3500

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

File: man/adoc/bpftrace.adoc
Comment (viktormalik) [inline man/adoc/bpftrace.adoc:2856]: The probe is triggered only when the **bpftrace process** receives a signal, right? Perhaps it's worth mentioning here to prevent confusion.

File: src/ast/attachpoint_parser.cpp
Comment (viktormalik) [inline src/ast/attachpoint_parser.cpp:293]: We could even enable wildcard attachments in future.

File: src/ast/passes/semantic_analyser.cpp
Comment (viktormalik) [inline src/ast/passes/semantic_analyser.cpp:3162]: These should be caught by `AttachpointParser`, shouldn't they?

Comment (viktormalik) [inline src/ast/passes/semantic_analyser.cpp:3162]: Also, could we get unit tests for these errors?

File: src/bpftrace.cpp
Comment (viktormalik) [inline src/bpftrace.cpp:955]: It looks like it's time to start storing special probes in a map hashed by the probe name. That could remove these loops and simplify the code a lot.

**Review by viktormalik — COMMENTED** [review]
Nice feature! Just a few comments below.

**Comment by ajor** [comment]
To throw out another idea for probe naming that combines two cases we'd eventually want to support:
```
trigger:signal:SIGUSR1 {}
trigger:keypress:space {}
```

File: src/ast/passes/semantic_analyser.cpp
Comment (amiremohamadi) [reply src/ast/passes/semantic_analyser.cpp:3162]: oh right, checking the number of arguments in AttachpointParser seems enough.

**Comment by amiremohamadi** [comment]
> To throw out another idea for probe naming that combines two cases we'd eventually want to support:
> 
> ```
> trigger:signal:SIGUSR1 {}
> trigger:keypress:space {}
> ```

looks more clear. if you all agree, we can use this naming.

**Comment by amiremohamadi** [comment]
or maybe using `event` instead of `trigger`?

```
event:signal:SIGUSR1 {}
```

**Comment by viktormalik** [comment]
What is the purpose of using the `trigger:` or `event:` prefix? To logically group probe types which fire when the bpftrace process is a target of an event (signal, keypress)? Are we expecting more such events in future?

**Comment by ajor** [comment]
> What is the purpose of using the `trigger:` or `event:` prefix? To logically group probe types which fire when the bpftrace process is a target of an event (signal, keypress)? Are we expecting more such events in future?

The logical grouping was the reason in my mind. I can't think of any classes of event like this right now other than signals and keypresses.

I guess another reason for grouping them behind a single probe type would be that it makes it clear that it's not tracing any signal or keypress systemwide - only ones sent to the bpftrace runtime. Another idea based on this:
```
bpftrace:signal:SIGUSR
bpftrace:keypress:space
```

File: src/bpftrace.cpp
Comment (amiremohamadi) [reply src/bpftrace.cpp:955]: how about using a function instead? if we store special probes in a map, in different parts across the code (E.g [this](https://github.com/bpftrace/bpftrace/blob/57b355c95b5406742edeb4b99330e2253703615b/src/bpfbytecode.cpp#L190)) we need to use templates or convert the std::map to a std::vec.

Comment (viktormalik) [reply src/bpftrace.cpp:955]: Sure, it was just an idea. We're talking units of probes here so storing them in a map will not bring any real performance benefits and, as you notice, complicate other code paths. Function sounds like a way to go, the only problem is that sometimes we need to *attach and trigger* a probe (BEGIN, END) and sometimes *just attach* it (signals).

File: src/bpftrace.cpp
Comment (viktormalik) [inline src/bpftrace.cpp:127]: Not valid ATM but once we support multiple signals, we'll need to add the `_<func>` part to the attach point name, right?

File: src/types.h
Comment (viktormalik) [inline src/types.h:600]: Shall we make a `bt` alias? It'd be unified with the common extension of bpfscript files.

**Review by viktormalik — COMMENTED** [review]
Nice changes! A few more comments but I think that it's almost ready.

**Comment by jordalgo** [comment]
Nit: The commit message could use some more detail and perhaps an example.

File: man/adoc/bpftrace.adoc
Comment (fbs) [inline man/adoc/bpftrace.adoc:2880]: does this need more words about how this overrides the built in behaviour (of printing maps) and how the bpf prog trigger is indirct (i.e. it depends on the runtime calling a trigger function). Also, what happens if you define two of these probes?

**Comment by jordalgo** [comment]
I don't mean to hold up this PR but I got confused when I first saw this naming `bpftrace:signal:SIGUSR`. I read the conversation above and understand the reasoning but naming a probe after the name of the tool feels wrong. I'm still thinking about alternatives but what about something like `self:signal:SIGUSR` ?

File: src/ast/attachpoint_parser.cpp
Comment (jordalgo) [inline src/ast/attachpoint_parser.cpp:293]: Eh, I'm not sure we should add wildcard support - and it definitely doesn't apply to BEGIN/END

Comment (viktormalik) [reply src/ast/attachpoint_parser.cpp:293]: True, I'm not sure either. It would make it possible to attach a single probe to all signals received by the bpftrace runtime but I don't know if that's any useful.

Once we add the `bpftrace:keypress:...` events, wildcards could be more useful.

**Comment by jordalgo** [comment]
Should this also deprecate the ability to print maps when issuing a SIGUSR to bpftrace? Seems strange to have both.
https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#map-printing

**Comment by amiremohamadi** [comment]
hmm. while defining a signal probe changes the default behavior, still it's not a breaking change for those who don’t use it. does it really need a migration guide entry?

**Comment by viktormalik** [comment]
> hmm. while defining a signal probe changes the default behavior, still it's not a breaking change for those who don’t use it. does it really need a migration guide entry? @viktormalik @jordalgo

But it is a breaking change for those who do. If people have scripts which run a bpftrace command, periodically send SIGUSR1, and process the results, this change will break them. Which is why I think that it is a breaking change.

**Comment by viktormalik** [comment]
Also cc @andreasgerstmayr who added the original "print maps on SIGUSR1" feature.

File: src/bpftrace.cpp
Comment (viktormalik) [inline src/bpftrace.cpp:1274]: I don't think that we should preserve the default behaviour here. It was just a convenience b/c some people wanted to inspect map contents during bpftrace script runtime but now that we have proper signal probes, they can just define it themselves.

I now realized that I confused you a lot b/c I missed this line in my earlier review and I though that we were dropping this behaviour. Which is why I insisted on the migration guide entry. My apologies about that.

Still, I think that we should drop it.

Comment (fbs) [reply src/bpftrace.cpp]: Isn't it nice to keep it? For simple scripts it can be a nice bit of utility, you dont have to add the `signal{print(@)}` manually, which makes it a lttle bit simpler, and the code complexity of keeping this seems low.

Comment (viktormalik) [reply src/bpftrace.cpp]: To me, it's a bit of a surprising behaviour, even though it is documented. Why should the tool print maps upon receiving some random signal? I understand that it was a convenient way to provide a requested feature but now that we have proper signal probes, I'd argue that we should stick to the principle of the least surprise.

File: src/ast/passes/semantic_analyser.cpp
Comment (danobi) [inline src/ast/passes/semantic_analyser.cpp:3273]: Looks like #3524 caused this duplicated code.

Will need a rebase past #3538. Let's do a clean fix to intent is saved into git.

**Comment by danobi** [comment]
Apologies in advance if this sends you on bit of a side quest, but BPF_PROG_TYPE_RAW_TRACEPOINT was added in 4.17 kernel: https://github.com/torvalds/linux/commit/a0fe3e574b50461c4811ce81811f0eaefda62871

bpftrace's policy w.r.t. old kernels is we support up to the oldest LTS kernel still in service: https://www.kernel.org/category/releases.html

The oldest LTS kernel still in service is 4.19, so we are free to assume BPF_PROG_TYPE_RAW_TRACEPOINT is always available.

So let's delete all the trigger probe stuff. It was always a bit of a hack. It should greatly simplify some of the stuff you're doing in this PR. Also means we can delete the feature probe (`has_raw_tp_special()`)

**Comment by amiremohamadi** [comment]
@danobi just to make sure i got your point correctly, now we don't need to attach the special probes to anyting? calling `bpf_prog_test_run_opts` would be enough, right?

File: src/bpftrace.h
Comment (amiremohamadi) [inline src/bpftrace.h:193]: using optional instead of initializing the `prog_fd` with a value because `bpf_prog_test_run_opts` will raise an error for errored fds (similiar to BEGIN/END)

**Comment by danobi** [comment]
> @danobi just to make sure i got your point correctly, now we don't need to attach the special probes to anyting? calling `bpf_prog_test_run_opts` would be enough, right?

Yes, that's right. And sorry about late reply - must've missed notification

**Comment by danobi** [comment]
> @viktormalik - Making sure you saw my comment above. What do you think about `self:signal:SIGUSR` instead of `bpftrace:signal:SIGUSR`?

Hmm, I bpftrace prefix is ok. Maybe I write too much c++/rust. In both languages you end up having a top level namespace w/ same name as project. `self` is probably ok too.

**Comment by danobi** [comment]
:P sure. Sounds like `self` is shorter and less ambiguous for future use cases.

`<joke>or we could do "this"</joke>`

**Comment by amiremohamadi** [comment]
thanks folks. I’d like to know @ajor and @viktormalik opinion too before doing any modifications.

**Comment by viktormalik** [comment]
Yeah, `self` sounds good and probably expresses the purpose better (the "target" is the current bpftrace process). No strong opinion on my side.

**Review by jordalgo — APPROVED** [review]
LGTM! Thanks for the contribution!

This does make me crave the ability for bpftrace to be more interactive and be able to handle custom commands from stdin e.g. "print_maps", "lost_event_count", "print_map_value(@a, 1)", etc.

File: src/ast/passes/semantic_analyser.cpp
Comment (amiremohamadi) [reply src/ast/passes/semantic_analyser.cpp:3339]: can you elaborate more? currently It looks like this 
https://github.com/bpftrace/bpftrace/blob/611450e2d362fd29543fafffefd81b05555ae39c/tests/semantic_analyser.cpp#L2197-L2200

**Comment by amiremohamadi** [comment]
> This does make me crave the ability for bpftrace to be more interactive and be able to handle custom commands from stdin e.g. "print_maps", "lost_event_count", "print_map_value(@A, 1)", etc.

if I’m not mistaken, we can also [receive data via signals](https://man7.org/linux/man-pages/man2/sigaction.2.html). while this approach isn’t as convenient as sending data through stdin, it can still be implemented relatively easily in the future

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [reply src/ast/passes/semantic_analyser.cpp:3339]: Don't mind my passing comment. It was just about the variable name "func" not really being a function. Not blocking anything.

**Comment by jordalgo** [comment]
@viktormalik @danobi - One last look before I merge?

File: src/required_resources.h
Comment (viktormalik) [inline src/required_resources.h:135]: Prefer `std::unordered_map` for string keys.

File: src/bpfbytecode.cpp
Comment (viktormalik) [inline src/bpfbytecode.cpp:191]: Not very nice that we have to copy here but I don't have a better solution at the moment.

**Comment by danobi** [comment]
Did a quick skim - didn't spot any major issues

**Comment by simpleton** [comment]
> BPF_PROG_TYPE_RAW_TRACEPOINT was added in 4.17 kernel

this's not sufficient, due to the test_run on raw_tracepoint is not available util 5.10 kernel. https://github.com/torvalds/linux/commit/1b4d60ec162f82ea29a2e7a907b5c6cc9f926321

**Comment by jordalgo** [comment]
@simpleton Sadly, our docs are a bit out of date. I think we're right now looking at a min version of 5.15 as our support backstop - though we're still investigating what's broken on that.

**Pull request merged**
